### PR TITLE
⚠️ Propagate context.Context throughout the codebase

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

/assign @detiber @alvaroaleman 
/milestone v0.7.x

If anyone has better ideas on how to propagate this context, please let me know 😄

### CODE REVIEW

File: pkg/internal/controller/controller.go
Comment (alvaroaleman) [inline pkg/internal/controller/controller.go:133]: I think `TODO` is better suited since we want to eventually change the `Runnable` signature to take a context and then use that context here

File: pkg/internal/controller/controller.go
Comment (detiber) [inline pkg/internal/controller/controller.go:179]: This can likely use `wait.UntilWithContext()` now

Comment (vincepri) [reply pkg/internal/controller/controller.go]: Done!

Comment (detiber) [reply pkg/internal/controller/controller.go]: :grimacing: maybe that will need to wait until after context is plumbed through Runnable, looks like test is not happy with the ordering of stopping leader election compared to the workers because of this.

File: pkg/manager/internal.go
Comment (vincepri) [inline pkg/manager/internal.go:135]: These have been removed in favor of `internalCtx/internalCancel`

File: pkg/manager/internal.go
Comment (vincepri) [inline pkg/manager/internal.go:185]: Going to add some documentation on what these are used for internally once we agree this is what we want / have some feedback

File: pkg/manager/internal.go
Comment (vincepri) [inline pkg/manager/internal.go:540]: The caller already defers this cancel, removed it from here

File: pkg/internal/controller/controller.go
Comment (vincepri) [inline pkg/internal/controller/controller.go:118]: Should we modify `Watch` to also accept a context, or save it as part of `Start` and pass it in here?

Comment (vincepri) [reply pkg/internal/controller/controller.go]: Propagated the context with the one passed into Start, given that seems the most consistent approach

File: pkg/manager/internal.go
Comment (vincepri) [reply pkg/manager/internal.go:185]: `Add` is where I hit a blocker, if we pass a context in there, we might not be able to cancel all running routines depending on which context is passed in. 

I had to go with an internal context and force it on every `Add` call to make sure that we cancel that context before exiting.

The other issue I found with having a tree of context structs is that when you do `WithCancel`, the cancel is propagated and we don't really want that for this use case.

I'm happy to iterate on it in multiple small steps, this was the one that was the least confusing 😄

File: pkg/controller/controller_integration_test.go
Comment (vincepri) [reply pkg/controller/controller_integration_test.go:83]: Removed the one on line 51, it was a leftover.

File: pkg/internal/controller/controller.go
Comment (vincepri) [reply pkg/internal/controller/controller.go:75]: Not that I could think about, the code today isn't really structured to use contexts how they're supposed to work, and I didn't want to break too much of the APIs yet, until we get more familiar with the internals.

`Watch()` is called both internally and externally, if we change the signature to pass a context as first parameter, we get into a situation where some watches might be using a context and some other are using another, which isn't ideal

File: pkg/manager/signals/signal.go
Comment (vincepri) [reply pkg/manager/signals/signal.go:30]: We pass this directly to `Start` in multiple places, which now accepts a context

File: pkg/manager/internal.go
Comment (alvaroaleman) [reply pkg/manager/internal.go:185]: Yeah I think this won't be possible without storing it, unless we want to create a new context on every `Add` and then store a slice of cancel funcs that are all called on stop, which seems a lot worse

File: pkg/manager/internal.go
Comment (alvaroaleman) [inline pkg/manager/internal.go:246]: Shoulnd't this get replaced with an `inject.ContextInto`?

File: pkg/manager/internal.go
Comment (alvaroaleman) [inline pkg/manager/internal.go:649]: If I didn't miss anything the `ctx` here is always `cm.internalCtx` so wouldn't it be easier to access that from the `controllerManager` rather than adding a new arg?

File: pkg/manager/signals/signal.go
Comment (alvaroaleman) [reply pkg/manager/signals/signal.go:30]: In a way jason has a good point though, I think the reason the context package has the `don't store it in a struct` comment is because a context is not only for canceling but also to transport request-scoped information. But we don't do that here, we only care about the canceling aspect, so the channel might actually a better fit than the context?

File: pkg/manager/internal.go
Comment (vincepri) [reply pkg/manager/internal.go:649]: Yeah good point

File: pkg/manager/internal.go
Comment (vincepri) [reply pkg/manager/internal.go:246]: It depends, given that we should really pass context as the first parameter, instead of injecting it into structs

File: pkg/manager/signals/signal.go
Comment (vincepri) [reply pkg/manager/signals/signal.go:30]: Yeah, wasn't objecting to it, just wanted to point out that we're accepting a context now. In a way, it works like stdlib behaves in a lot of places (http server is an example) so I copied that approach

**Comment by vincepri** [comment]
@DirectXMan12 Do you have some time to go over these changes? I'm happy to drop this PR if it's too many changes together and continue with smaller changes over time

File: pkg/manager/internal.go
Comment (alvaroaleman) [reply pkg/manager/internal.go:250]: Well the problem is that due to this terrible inject pattern the code will continue to compile but not work as before (I would assume it will just panic now for most ppl that relied on this). Can we maybe keep this for now, inject `cm.internalCtx.Done` and print a big fat deprecation warning in `inject.StopChannelInto` if this happens?

How is kubebuilder the CLI going to migrate projects?

File: pkg/manager/internal.go
Comment (alvaroaleman) [inline pkg/manager/internal.go:416]: could we change this to `func(_ context.Context` to make it clearer, that the context we pass here is not used?

File: pkg/manager/manager_test.go
Comment (alvaroaleman) [inline pkg/manager/manager_test.go:58]: Rather than creating and canceling a context in each testcase, can we just create the ctx and cancel in the BeforeEach and `cancel` in the `AfterEach`?

File: pkg/runtime/inject/inject.go
Comment (alvaroaleman) [inline pkg/runtime/inject/inject.go:106]: As mentioned elsewhere, I don't think we should just remove it, because that will result in hard to debug panics down the line.

File: pkg/manager/manager_test.go
Comment (vincepri) [reply pkg/manager/manager_test.go:58]: The race detector will complain about using the context concurrently, so I changed it that way

File: pkg/runtime/inject/inject.go
Comment (vincepri) [reply pkg/runtime/inject/inject.go:106]: I'd rather have the context properly propagated within reconciler calls, rather than having depedency injection like these around, which would result in the context to be stored as a struct field

File: pkg/manager/internal.go
Comment (vincepri) [reply pkg/manager/internal.go:250]: I see, would we pass the ctx.Done() channel here then?

File: pkg/runtime/inject/inject.go
Comment (vincepri) [reply pkg/runtime/inject/inject.go:106]: (See reply above, we can inject ctx.Done())

File: pkg/manager/internal.go
Comment (joelanford) [reply pkg/manager/internal.go:250]: > Can we maybe keep this for now, inject cm.internalCtx.Done and print a big fat deprecation warning in inject.StopChannelInto if this happens?

:+1: 

> How is kubebuilder the CLI going to migrate projects?

Sounds like we'll keep this interface and injection in for now, so no changes necessary on the kubebuilder scaffolding side related to this particular aspect of the PR?

File: pkg/internal/controller/controller.go
Comment (vincepri) [reply pkg/internal/controller/controller.go:76]: Added a comment around the use of a context in struct

File: pkg/manager/signals/signal.go
Comment (vincepri) [reply pkg/manager/signals/signal.go]: Fixed, thank you all

File: pkg/manager/internal.go
Comment (vincepri) [reply pkg/manager/internal.go:250]: @alvaroaleman Mind reviewing the last changes related to `StopChannelInto`, I'm a little confused about this interface and the whole injection code left me with some questions, but we can fix those later. For the purpose of this PR, can you double check that the logic here is restored as you'd expect?

File: pkg/manager/internal.go
Comment (vincepri) [reply pkg/manager/internal.go]: Done

File: pkg/controller/controller_integration_test.go
Comment (vincepri) [reply pkg/controller/controller_integration_test.go]: It should be used now, thank you!

File: pkg/controller/controller.go
Comment (vincepri) [reply pkg/controller/controller.go:68]: Done, thank you!

File: pkg/cache/informertest/fake_cache.go
Comment (vincepri) [reply pkg/cache/informertest/fake_cache.go:125]: Do you want to do this in tests as well?

File: pkg/cache/multi_namespace_cache.go
Comment (vincepri) [reply pkg/cache/multi_namespace_cache.go:97]: Did a quick scan throughout our codebase and fixed all the comments I could find that were outdated (e.g. referenced a stop channel)

File: pkg/envtest/webhook_test.go
Comment (vincepri) [reply pkg/envtest/webhook_test.go:37]: Done!

File: pkg/cache/internal/deleg_map.go
Comment (vincepri) [reply pkg/cache/internal/deleg_map.go:64]: Actually, just reverted this to still return nil, I'll open a different issue to tackle this in a different PR, but all the tests are expecting Start to return nil, and it doesn't take into account `context.Cancelled` error when the context is cancelled at the end of a test, does that sound good?

File: pkg/manager/internal.go
Comment (alvaroaleman) [reply pkg/manager/internal.go:250]: I don't see anything related to that, it still seems to be gone, am I completely missing it?

File: pkg/cache/internal/deleg_map.go
Comment (alvaroaleman) [reply pkg/cache/internal/deleg_map.go:64]: I think that makes more sense, because usually the manager is started with something like
```
if err := mgr.Start(ctx); err != nil {
    log.Fatal(err, "manager failed")
}
```

even thought a canceled context is an expected condition.

File: pkg/manager/internal.go
Comment (vincepri) [reply pkg/manager/internal.go:250]: Done, must have forgotten to push the changes

Comment (alvaroaleman) [reply pkg/manager/internal.go:250]: I still only see the removed code?

Comment (vincepri) [reply pkg/manager/internal.go:250]: * actually pushed now *

File: pkg/cache/cache_test.go
Comment (alvaroaleman) [inline pkg/cache/cache_test.go:398]: Why do we need a new `context` here and in the two tests below?

File: pkg/manager/manager_test.go
Comment (alvaroaleman) [inline pkg/manager/manager_test.go:1247]: I think we should keep that or not?

**Review by alvaroaleman — COMMENTED** [review]
Two questions, other than that lgtm. It would be great if we could fine someone else to give this a pass though in case I missed something due to the sheer size of the PR.

File: pkg/cache/cache_test.go
Comment (vincepri) [reply pkg/cache/cache_test.go]: Fixed

File: pkg/manager/manager_test.go
Comment (vincepri) [reply pkg/manager/manager_test.go:1247]: Fixed

**Comment by vincepri** [comment]
@DirectXMan12 @ncdc @detiber If you have some time, an extra set of 👀 would be greatly appreciated

**Review by alvaroaleman — APPROVED** [review]
/lgtm
/hold
so others can have a look

**Comment by ncdc** [comment]
I think this LGTM. There's an open question still for FakeInformers.Start() - ?

**Comment by vincepri** [comment]
Thanks! I'm ok having that come in later, it might break tests if we return ctx.Err(), which seems we've decided not to do in https://github.com/kubernetes-sigs/controller-runtime/pull/1116#discussion_r493008378

**Comment by vincepri** [comment]
/hold cancel

Checked with @alvaroaleman before removing the hold, let's get this merged so we can unblock other work :)

**Pull request merged**
